From 21c5db303b81d34b28ca83e46da1eca079bd84a3 Mon Sep 17 00:00:00 2001
From: Your Name <you@example.com>
Date: Fri, 15 Mar 2024 07:52:40 -0300
Subject: [PATCH] Change Makefile to be compatible with yocto build

---
 Makefile | 14 +++++++-------
 1 file changed, 7 insertions(+), 7 deletions(-)

diff --git a/Makefile b/Makefile
index eada865..1e57a67 100755
--- a/Makefile
+++ b/Makefile
@@ -25,8 +25,6 @@ CC =		$(CROSS_COMPILE)gcc
 endif
 
 LD ?=		$(CROSS_COMPILE)ld
-BACKUP=		/root/backup
-YMD=		`date +%Y%m%d%H%M`
 
 #############################################################################
 # Configuration Options
@@ -121,7 +119,6 @@ endif
 
 
 KERNELVERSION_X86 := 	$(shell uname -r)
-KERNELDIR ?= /lib/modules/$(KERNELVERSION_X86)/build
 LD += -S
 
 
@@ -133,7 +130,7 @@ APPDIR= $(shell if test -d "mapp"; then echo mapp; fi)
 # Compiler Flags
 #############################################################################
 
-	ccflags-y += -I$(KERNELDIR)/include
+	ccflags-y += -I$(KERNEL_SRC)/include
 
 	ccflags-y += -DFPNUM='"88"'
 
@@ -218,7 +215,7 @@ endif
 
 
 # add -Wno-packed-bitfield-compat when GCC version greater than 4.4
-GCC_VERSION := $(shell echo `gcc -dumpversion | cut -f1-2 -d.` \>= 4.4 | sed -e 's/\./*100+/g' | bc )
+GCC_VERSION := $(shell echo `${CC} -dumpversion | cut -f1-2 -d.` \>= 4.4 | sed -e 's/\./*100+/g' | bc )
 ifeq ($(GCC_VERSION),1)
 	ccflags-y += -Wno-packed-bitfield-compat
 endif
@@ -473,13 +470,13 @@ pciexxx-objs := $(MOALOBJS)
 else
 
 default:
-	$(MAKE) -C $(KERNELDIR) M=$(PWD) modules
+	$(MAKE) -C $(KERNEL_SRC) M=$(PWD) modules
 
 endif
 
 ###############################################################
 
-export		CC LD ccflags-y KERNELDIR
+export		CC LD ccflags-y KERNEL_SRC 
 
 ifeq ($(CONFIG_STA_SUPPORT),y)
 ifeq ($(CONFIG_UAP_SUPPORT),y)
@@ -671,4 +668,7 @@ endif
 	$(MAKE) -C mapp/mlanevent $@
 endif
 
+modules_install:
+	$(MAKE) -C $(KERNEL_SRC) M=$(PWD) modules_install
+
 # End of file
-- 
2.25.1

