diff --git a/Makefile b/Makefile
index 8935a20..d0322ab 100755
--- a/Makefile
+++ b/Makefile
@@ -32,10 +32,6 @@ else
 CC ?=		$(CROSS_COMPILE)gcc -I$(COMPATDIR)
 endif
 
-LD ?=		$(CROSS_COMPILE)ld
-BACKUP=		/root/backup
-YMD=		`date +%Y%m%d%H%M`
-
 #############################################################################
 # Configuration Options
 #############################################################################
@@ -159,7 +155,6 @@ ccflags-y += -DIMX_ANDROID
 CONFIG_ANDROID_KERNEL=y
 endif
 endif
-KERNELDIR ?= /usr/src/arm/linux_5_10_y_kernel/linux-nxp
 CROSS_COMPILE ?= /opt/fsl-imx-internal-xwayland/5.10-gatesgarth/sysroots/x86_64-pokysdk-linux/usr/bin/aarch64-poky-linux/aarch64-poky-linux-
 
 LD += -S
@@ -171,8 +166,6 @@ APPDIR= $(shell if test -d "mapp"; then echo mapp; fi)
 # Compiler Flags
 #############################################################################
 
-	ccflags-y += -I$(KERNELDIR)/include
-
 	ccflags-y += -DFPNUM='"92"'
 
 ifeq ($(CONFIG_DEBUG),1)
@@ -645,13 +638,13 @@ moal-objs := $(MOALOBJS)
 else
 
 default:
-	$(MAKE) -C $(KERNELDIR) M=$(PWD) ARCH=$(ARCH) CROSS_COMPILE=$(CROSS_COMPILE) modules
+	$(MAKE) -C $(KERNEL_SRC) M=$(PWD) ARCH=$(ARCH) CROSS_COMPILE=$(CROSS_COMPILE) modules
 
 endif
 
 ###############################################################
 
-export		CC LD ccflags-y KERNELDIR
+export		CC LD ccflags-y KERNEL_SRC
 
 ifeq ($(CONFIG_STA_SUPPORT),y)
 ifeq ($(CONFIG_UAP_SUPPORT),y)
